ASoC: Intel: soc-acpi-intel-ptl-match: Add support for rt722_l0_rt132…

macchian opened the pull request:
…0_l23

This patch adds support for the RT722 codec on SoundWire link 0, and RT1320 left and right amplifiers on links 2 and 3, respectively.

The RT722 speaker function is excluded, as it is the aggregate configuration.

bardliao commented on sound/soc/intel/common/soc-acpi-intel-ptl-match.c:
```diff
@@ -735,6 +794,7 @@ struct snd_soc_acpi_mach snd_soc_acpi_intel_ptl_sdw_machines[] = {
 		.sof_tplg_filename = "sof-ptl-rt722.tplg",
 		.get_function_tplg_files = sof_sdw_get_tplg_files,
 	},
+
```
unnecessary new line

bardliao commented:
@macchian My main concern is that no speaker endpoint is included in the rt722's `.endpoints`. It will not work if the rt722 speaker + 2 * rt1320 config is used in the future. In theory, we should add all endpoints to the `.endpoints` and the driver will skip the endpoint that is not present in the DisCo table. Or use a DMI quirk to disable the rt722's speaker. Note that you can't use the kernel with https://git.kernel.org/pub/scm/linux/kernel/git/broonie/sound.git/commit/?id=4f8ef33dd44a3d1136d393 if the BIOS doesn't provide the proper DisCo table.

macchian committed:
ASoC: Intel: soc-acpi-intel-ptl-match: Add support for rt722_l0_rt1320_l23

This patch adds support for the RT722 codec on SoundWire link 0,
and RT1320 left and right amplifiers on links 2 and 3, respectively.

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

macchian committed:
ASoC: Intel: sof_sdw: Add quirk to exclude RT722 speaker

This SKU does not use the RT722 internal speaker. Add
SOC_SDW_CODEC_SPKR to prevent DAI being created via
codec_info_list.dais.quirk_exclude.

Other SKUs continue to check higher-priority sdca_functions
when the BIOS SDCA property is present.

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

bardliao commented on sound/soc/intel/boards/sof_sdw_common.h:
```diff
@@ -46,11 +46,11 @@ enum {
 #define SOF_BT_OFFLOAD_SSP(quirk)	\
 	(((quirk) << SOF_BT_OFFLOAD_SSP_SHIFT) & SOF_BT_OFFLOAD_SSP_MASK)
-#define SOF_SSP_BT_OFFLOAD_PRESENT	BIT(18)
+#define SOF_SSP_BT_OFFLOAD_PRESENT	BIT(21)
```
Good catch. Not sure if it needs a `fixup` tag. Also, it will impact the devices that use module parameter to overwrite the quirk. @ujfalusi @kv2019i How should we handle it?

macchian replied:
@kv2019i @ujfalusi , I added a PR to support bt_link_mask starting from ACE4+ based on @bardliao PR #5518, for your review.

kv2019i replied:
I think this is @bardliao ok as we have not upstreamed the BT offload quirk bits yet. All shipping devices (all chromebooks) have a hardcoded SSP port for BT, so I think this is ok change to make.

kv2019i replied:
Ok update. So we have had the SOF_BT_OFFLOAD* defs in upstream for some time, and as all flags are exposed via the quirk module param, these have been available as module params as well. 

So this is a bit iffy situation (created when common flags were moved to vendor neutral soc_sdw_utils.h). 

There is a conflict, so we need to resolve this anyways. I think this PR is the right path. The BT_OFFLOAD bits have only been used in the DMI overrides and given this feature is not availble outside Chromebooks, there's no known usage of this by end-users (via module params). The SOC_SDW_CODEC_SPKR bit OTOH has seen a lot of use by actual users, so this we cannot change at this point.

So as a summary, I think the change in this PR is good. We do need to be very careful aligning soc_sdw_utils.h and this SOF specific file.

macchian committed:
ASoC: Intel: sof_sdw: create SSP BT DAI if bt_link_mask sets

This patch adds support for selecting the SSP DAI for Bluetooth offload
based on the machine parameter `bt_link_mask` when provided and on ACE4
or newer Intel SoCs. This allows platform data to override which SSP
port is used for BT offload topologies, typically named:

    sof_tplg_filename"-%dch"-ssp%d-bt.tplg

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

macchian committed:
fixup! ASoC: Intel: sof_sdw: shift SSP BT mask bits.

The SSP BT mask bits overlapped with SOC_SDW_CODEC_SPKR,
SOC_SDW_SIDECAR_AMPS, and SOC_SDW_CODEC_MIC BIT[15–17] in
sdw_utils.h. Shift the SSP BT mask bits to a higher range to
eliminate the conflict.

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

bardliao commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1143,8 +1145,28 @@ static int create_bt_dailinks(struct snd_soc_card *card,
+	 * and if the SoC hardware IP version is ACE4 or newer.
+	 * This allows platform data to override which SSP is used
+	 * for Bluetooth offload.
+	 */
```
`sof_sdw_quirk` could also from module parameter. Not sure which one should have higher priority.

macchian replied:
Even when the sof_sdw_quirk override is used alone or together with bt_link_mask, it is the bt_link_mask that ultimately determines which topology with the BT prefix is loaded.
